std/container/rbtree.d

```diff
@@ -1724,10 +1724,82 @@ assert(equal(rbt[], [5]));
      * Constructor. Pass in an array of elements, or individual elements to
      * initialize the tree with.
      */
-    this(Elem[] elems...)
+    this(scope Elem[] elems...)
     {
         _setup();
-        stableInsert(elems);
+        foreach (e; elems)
```

Comment:
Did you just copy and paste the code from `stableInsert` into the function?

If this is necessary to get DIP1000 to compile, then something is very wrong.

Reply:
@JackStouffer I was about to ask the same question. Why can't we also instrument `stableInsert` and `_add`? I suppose `stableInsert` should auto-detect scope-ness, but `_add` is not a template.

Reply:
Frankly, I did expect exactly this question and I know, the fix is ugly, but:
The main problem was, that
private auto _add(Elem n)
returns a Node or a Tuple containing a Node, which raised the error:
scope variable result may not be returned (the same error, that I originally encountered when fixing slist.d).
Thus I would have to change _add or include a new overload for _add or inline a changed version.
In principle, it's  copy and paste the code from stableInsert and _add, as well as the changes to _add to "virtually" return void.
Btw, the return of _add's Node is nowhere used, only the added member.
The upside of this uglyness is avoiding to copy the elems passed to ctor

Reply:
I'm pretty confident in saying that the wrong answer in any situation is code duplication, especially since you'll have to do the same thing to stable insert itself eventually.

The right answer here is to rewrite `_add` return `size_t`, because as far as I can tell, the returned `Node` is not actually used anywhere.

Reply:
> The upside of this uglyness is avoiding to copy the elems passed to ctor

How so? `stableInsert(someArray)` should not copy anything.

Reply:
> I don't know why it's this way

Answer: I was smoking something when I did it in the first commit! Clearly, I thought it was needed, maybe for an interim version that used it for something, but then was pruned out.

Reply:
There is no dissent about code duplication.
Avoiding code duplication just didn't work in the slist.d case (but maybe for other reasons?); that's why I selected it as first approach.
Anyway, I'll try once more here.

Reply:
If DIP1000 _required_ code duplication in some other case, then DIP1000 is flawed.

Reply:
So how about changing the return value of `_add`, and making its parameter `scope`, so we can pull this and move on?

Reply:
because that's not so easy:

```
std/container/rbtree.d(1839):        instantiated from here: RedBlackTree!(string, "a < b", true)
std/container/rbtree.d(1913):        instantiated from here: redBlackTree!(true, string)
std/container/rbtree.d(940): Error: scope variable result may not be returned
```

and that's just for one of the two paths:

https://github.com/wilzbach/phobos/blob/012395e4dcb95b3c09b9e5d04d33871963c6b65d/std/container/rbtree.d#L934-L953

`Tuple` isn't `scope` either.

Reply:
Your proposal is basically the same what I tried already for std.container.slist (https://github.com/dlang/phobos/pull/6295; there it is function insertFront that I duplicated in the ctor to get it to compile) and it didn't work with master, but acc. to @wilzbach analysis with 2.079.
6295 got reverted therefore by https://github.com/dlang/phobos/pull/6353 and I think for a good reason (avoid code dupication).
I didn't yet try here (for _add) because of probably the same underlying problem for both slist.d and rbtree.d

Reply:
Oh and if we do `scope _add(scope Elem n)`, we have to deal with `_add`'s style of assigning to `this`:

```
std/container/rbtree.d(887): Error: scope variable result assigned to non-scope this._begin
std/container/rbtree.d(903): Error: scope variable result assigned to non-scope parameter newNode calling std.container.rbtree.RBNode!string.RBNode.left
std/container/rbtree.d(924): Error: scope variable result assigned to non-scope parameter newNode calling std.container.rbtree.RBNode!string.RBNode.right
std/container/rbtree.d(953): Error: scope variable result assigned to non-scope parameter _param_1 calling std.typecons.Tuple!(bool, "added", RBNode!string*, "n").Tuple.this
```

Reply:
For reference,  the closest I got was this - it "only" requires two lines of wrongly commenting out:

<details>

```diff
diff --git a/std/container/rbtree.d b/std/container/rbtree.d
index a8945f082..c17337e70 100644
--- a/std/container/rbtree.d
+++ b/std/container/rbtree.d
@@ -937,7 +937,7 @@ if (is(typeof(binaryFun!less(T.init, T.init))))
             debug(RBDoChecks)
                 check();
             ++_length;
-            return result;
+            return result is null;
         }
         else
         {
@@ -950,7 +950,7 @@ if (is(typeof(binaryFun!less(T.init, T.init))))
             }
             debug(RBDoChecks)
                 check();
-            return Tuple!(bool, "added", Node, "n")(added, result);
+            return added;
         }
     }
 
@@ -1122,7 +1122,7 @@ if (is(typeof(binaryFun!less(T.init, T.init))))
      *
      * Complexity: $(BIGOH log(n))
      */
-    size_t stableInsert(Stuff)(Stuff stuff) if (isImplicitlyConvertible!(Stuff, Elem))
+    size_t stableInsert(Stuff)(scope Stuff stuff) @safe if (isImplicitlyConvertible!(Stuff, Elem))
     {
         static if (allowDuplicates)
         {
@@ -1131,7 +1131,7 @@ if (is(typeof(binaryFun!less(T.init, T.init))))
         }
         else
         {
-            return(_add(stuff).added ? 1 : 0);
+            return _add(stuff) ? 1 : 0;
         }
     }
 
@@ -1143,7 +1143,7 @@ if (is(typeof(binaryFun!less(T.init, T.init))))
      *
      * Complexity: $(BIGOH m * log(n))
      */
-    size_t stableInsert(Stuff)(Stuff stuff) if (isInputRange!Stuff && isImplicitlyConvertible!(ElementType!Stuff, Elem))
+    size_t stableInsert(Stuff)(scope Stuff stuff) @safe if (isInputRange!Stuff && isImplicitlyConvertible!(ElementType!Stuff, Elem))
     {
         size_t result = 0;
         static if (allowDuplicates)
@@ -1158,8 +1158,8 @@ if (is(typeof(binaryFun!less(T.init, T.init))))
         {
             foreach (e; stuff)
             {
-                if (_add(e).added)
-                    ++result;
+                //if (_add(e))
+                    //++result;
             }
         }
         return result;
@@ -1401,7 +1401,7 @@ assert(equal(rbt[], [5]));
     }
 
     /++ Ditto +/
-    size_t removeKey(U)(U[] elems)
+    size_t removeKey(U)(scope U[] elems)
         if (isImplicitlyConvertible!(U, Elem))
     {
         immutable lenBefore = length;
@@ -1423,7 +1423,7 @@ assert(equal(rbt[], [5]));
     }
 
     /++ Ditto +/
-    size_t removeKey(Stuff)(Stuff stuff)
+    size_t removeKey(Stuff)(scope Stuff stuff)
         if (isInputRange!Stuff &&
            isImplicitlyConvertible!(ElementType!Stuff, Elem) &&
            !isDynamicArray!Stuff)
@@ -1724,7 +1724,7 @@ assert(equal(rbt[], [5]));
      * Constructor. Pass in an array of elements, or individual elements to
      * initialize the tree with.
      */
-    this(Elem[] elems...)
+    this(scope Elem[] elems...)
     {
         _setup();
         stableInsert(elems);
@@ -1733,7 +1733,7 @@ assert(equal(rbt[], [5]));
     /**
      * Constructor. Pass in a range of elements to initialize the tree with.
      */
-    this(Stuff)(Stuff stuff) if (isInputRange!Stuff && isImplicitlyConvertible!(ElementType!Stuff, Elem))
+    this(Stuff)(scope Stuff stuff) if (isInputRange!Stuff && isImplicitlyConvertible!(ElementType!Stuff, Elem))
     {
         _setup();
         stableInsert(stuff);
```

</details>

Reply:
I decided to submit this patch as new PR, s.t. it's easier to look at it and improve upon it: https://github.com/dlang/phobos/pull/6362